src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Minimal.cs

```diff
@@ -265,7 +265,7 @@ public void CopyTo(int sourceIndex, T[] destination, int destinationIndex, int l
         public ImmutableArray<T>.Builder ToBuilder()
         {
             var self = this;
-            if (self.Length == 0)
+            if (self.IsEmpty)
```

Comment:
These changes have potential to introduce performance regression due to shared generic code inlining limitations in .NET Framework. I think you would want to manually inline `ImmutableArray.Length` into `ImmutableArray.IsEmpty` to make sure that the new code is as good as the old one.

Reply:
`ImmutableArray<T>.IsEmpty` already inlines `Length == 0`; is that what you are referring to by "manually inline"?

https://github.com/dotnet/runtime/blob/43d558387cc0d48e45f9c1ca960f912c425c1fa1/src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.Minimal.cs#L156-L160

Reply:
Take a look at: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4aNYgGYGuDFCFgMDALIBPAJL58QjLwA2MAIJQo2HQB4AKgD4aAbxoNPDBwG0AugzsNjqy1F4MHl6KDOJiDFwwIgDmGCwMALxOgda2fAnJqaHh0cAQEKYMergAovgcGDoZWfkpaenpDHRFUUql5ZU1dQ2kTdnBeYmtGR1dNAC+ctQ+AFLiGADiiTBQ4mAAFA0cMBAAZnvE5HQAlFd+8koXSAxkDADCjlnuYT3P5E+xmgcMBUFj2ugMRhMwHMVmCH0CVwYX3CKOIAHYGHtsBMCm0ZoiAPydBggBhwcjdTwLaiRTzRR4xOJAlQAITB+kMxjMlhy9mcCKRtJRzwx2KqtXqjSJjFJ5MpDGpQvpf0ZgOBGFe7IhXOhPLh/OwiORwrpor44qGOhG0pJZIpQupcyAA

TestA shows the code that people will get today.
TestB shows the significantly slower code that people will get with this change.
TestC shows how you need to change IsEmpty property to fix this regression

Reply:
Cool; seems to me that that is the only place that needs to be fixed.
Thanks for the suggestion.

Btw, it means that until now, `IsEmpty` was performing worse than `Length == 0` for `ImmutableArray<T>`?

Reply:
Yep, in certain situations.

It suggests that we may want to exclude `ImmutableArray<T>` from this analyzer/fixer. Switching from Count to IsEmpty does not provide any benefit, and it can be even actively harmful for existing versions of the package.

Reply:
> Is there any negative repercussions given the NonVersionable attribute?

NonVersionable attribute says that the method behavior cannot change in any way. It is ok to change the implementation in compatible way.

Reply:
> NonVersionable attribute says that the method behavior cannot change in any way. It is ok to change the implementation in compatible way.

What's a non-compatible way?

> It suggests that we may want to exclude ImmutableArray<T> from this analyzer/fixer. 

We already excluded Span/Memory from the rule given that in some cases, using Length performs better; this makes me wonder, what was the point of having an `IsEmpty` property in all these types if there is no benefit in using it?

Reply:
> What's a non-compatible way?

For example, changing the methods to not throw NullReferenceException when the array is null would be incompatible with `NonVersionableAttribute`.

> what was the point of having an IsEmpty property in all these types if there is no benefit in using it?

Syntax sugar. We have many small convenience methods in the framework that exist just to make the code look better. Ideally, these methods would have same performance as the more verbose implementations. Unfortunately, it is not always the case due to JIT limitations and bugs.

Reply:
> It suggests that we may want to exclude `ImmutableArray<T>` from this analyzer/fixer.

That sounds like the right thing to me.

Reply:
Or maybe at this point, we go back to just flagging specific collection types where we know it's a perf win.

Reply:
That's a good point, @stephentoub.  As often as we've seen this just within our own types, it's likely that other types in the ecosystem should not have this guidance applied to them, and typically the consumer of the types wouldn't go through the work of analyzing the perf and they'd just follow this analyzer's guidance.

@Jozkee -- I think you have the original list of types that we were confident this would help with, right?  Can you post the list of types where we know this would be a benefit here and we can decide if that's the approach we should take.

Reply:
>Or maybe at this point, we go back to just flagging specific collection types where we know it's a perf win.

I haven't analyzed the assembly size for all collection types but as of now based on early discussions I think that only encompasses concurrent collections.

Also, I don't think we should add ImmutableArray<T> to the exclusion list since there is already chance to optimize `IsEmpty`, instead we should be more careful with adding `IsEmpty` properties in the future if there is no perf. win at all.

Reply:
Thanks, @Jozkee.  That works for me here.  We should keep an eye on this to see if we keep finding more types that need special attention, especially once this analyzer ships with the SDK.

Reply:
> I don't think we should add ImmutableArray to the exclusion list 

Does the fixed understand the version of ImmutableArray that the project is targeting?

Reply:
@jkotas no, it does not know about which version is using. 
I assume your comment aims to the case where someone uses the analyzer in an earlier version (e.g. 3.1) where they will get the suggested fix that causes perf. decrease, right?

So what should we do? Here are the possible course of actions:

1. Exclude the type.
2. Flip over to an allow list for types as Stephen suggests.
3. Fix the perf. pitfall on this type and keep it included.

Reply:
My vote would be allow list as well.

Reply:
Also tagging @sharwell, who has been in similar discussions in the past and has taken the opposing view, but we haven't had a conclusive decision. Switching to an allow list would drastically reduce the scope of the analyzer, and also push up the maintainability (historically, analyzers based on allow lists tend to get more user feedback when they fire on some types and not on others, and it is difficult to identify if a type was excluded from the allow list on intent or by accident). My personal preference here would be 3, which if not possible then 1. 

> Does the analyzer/fixer understand the version of ImmutableArray that the project is targeting?

If required, this can be identified in the analyzer. It has access to the compilation and all referenced assemblies, so should be able to identify the version of the referenced `System.Collections.Immutable`.

Reply:
@Jozkee and I discussed this more and l considered it from several angles.  The XML scenario [seen above](https://github.com/dotnet/runtime/pull/39890#discussion_r459966637) worries me thinking about how much code in the wild could have the pattern of including an `IsEmpty` property as well as a `Count` or `Length`, where the properties do not mean the same as what we assume in the analyzer.  Doing a rudimentary search on GitHub surfaces thousands of cases where `IsEmpty` properties are defined and just spot-checking them we saw some cases where it's questionable whether or not this rule should apply. One example was a `Rect` class that doesn't quite match this pattern, but easily could have--that was particularly bothersome.

Even though the allow list could prompt more feedback from users, that seems like it would be relatively tame feedback compared to the scenario of the analyzer recommending changes that alter the meaning of the code in domain-specific classes (even like a `Vehicle` type we saw).

Here's my recommendation:

1. Switch to an allow list (referring back to our previous candidate list)
2. Apply the perf pitfall fix in `ImmutableArray<T>` ([as seen here](https://github.com/dotnet/runtime/pull/39890/commits/3277112a4f050f91cc0ce212de49926c28cabf9a))
3. If it can be done in an afternoon or so, conditionally include `ImmutableArray<T>` in that list only for .NET 5.0+

Unless there are any strong objections, we'll move forward with this in time for RC1.

Here are some interesting examples for reference:
* https://github.com/edisontung/playscript-mono/blob/87f9d7c0f81741d0076c70f22ce2de73f872c53e/mcs/class/System/System.Text.RegularExpressions/interval.cs#L79
* https://github.com/akil03/Kwizz/blob/bce8b10a14bd6793eaf32eb28566826c2018fb9e/Assets/UI/DoozyUI/Scripts/ScriptableObjects/NamesDatabase.cs
* https://github.com/Mauru/MauronAlpha.Csharp/blob/eb8303e248c459795501571a53804fc3d06dbf03/MauronAlpha.TextProcessing/Units/Word.cs

Reply:
I would keep it simple and kept `ImmutableArray<T>` excluded. Switching `Count == 0` to `IsEmpty` does not provide any performance benefit for `ImmutableArray<T>`, even after the fix for the perf pitfall.

Reply:
> Switch to an allow list (referring back to our previous candidate list)

@jeffhandley Would it be reasonable to take a middle ground for the proposal so that allow list includes all types that implement one or more of the following interfaces: `ICollection`, `ICollection<T>` or `IReadOnlyCollection`? I doubt if any collection type implementing these interfaces would have `IsEmpty` property that is not semantically equivalent to `Count == 0`.

Reply:
FWIW, [`XmlQueryType`](https://github.com/dotnet/runtime/pull/39890#discussion_r459966637) implements `ICollection<T>` indirecrly. 

So, if we add `ICollection`, `ICollection<T>` or `IReadOnlyCollection<T>` to the allow list, we would still have https://github.com/dotnet/runtime/pull/39890#discussion_r459966637 as a false positive.

Reply:
IMO a type implementing any collection type such that `.IsEmpty` returns a different result then `.Count == 0` has a _serious_ API bug that deserves a suppression in source (for not well known types) OR explicit addition to analyzer's excluded list (for well-known runtime types with this issue) in addition to an API bug filed on the API owner.

Any case, if we are really concerned about semantics change by going from Count check to IsEmpty check for any possible code out there, going conservative with a much smaller explicit set of allowed types is the only option.

Reply:
Thanks, @mavasani. I agree from a library perspective, and the Xml example here is indeed a misleading API. With `IsEmpty` being loose without any interfaces that define its semantics, I do think the conservative approach is appropriate here.  If there was an interface that included `IsEmpty` and we could limit to `ICollection` types that also use the interface defining `IsEmpty`, I think we'd be able to do this globally.

Reply:
> without any interfaces that define its semantics,

There's a nice quote for that:

```
I often hear people saying that interfaces specify
contracts. I believe this is a dangerous myth. Interfaces, by themselves,
do not specify much beyond the syntax required to use an object.

-Krzysztof Cwalina (Annotation), Framework Design Guidelines
```

That said, I think it's reasonable to say that ICollection being in the type's hierarchy does give a strong expectation to an `IsEmpty`; so perhaps the configuration option approach that Manish suggested would be either "known good types" or "any ICollection types" and maybe a third mode of "any type with both IsEmpty and Count".